cache/redis_cache_test.go

```diff
@@ -170,6 +170,56 @@ func TestKeyExpirationWhileFetchtingRedis(t *testing.T) {
 		t.Fatalf("expecting an error of type RedisCacheError, err=%s", err)
 	}
 }
+
+func TestCorruptedPayloadWhileFetchtingRedis(t *testing.T) {
+	cache, redis := getRedisCacheAndServer(t)
+	defer cache.Close()
+
+	key := &Key{
+		Query: []byte(fmt.Sprintf("SELECT test")),
+	}
+	payloadSize := 4 * 1024 * 1024
+	exepctedValue := strings.Repeat("a", payloadSize)
+	reader := strings.NewReader(exepctedValue)
+
+	if _, err := cache.Put(reader, ContentMetadata{Encoding: "ce", Type: "ct", Length: int64(payloadSize)}, key); err != nil {
+		t.Fatalf("failed to put it to cache: %s", err)
+	}
+
+	//simulating a data alteration
+	redisData, err := redis.Get(key.String())
+	if err != nil {
+		t.Fatalf("item not inseted in redis")
+	}
+
+	err = redis.Set(key.String(), redisData[:payloadSize-100])
+	if err != nil {
+		t.Fatalf("could not alter data in redis")
+	}
+	//simulate a value almost expired
+	redis.SetTTL(key.String(), 2*time.Second)
+
+	_, err = cache.Get(key)
+	_, isRedisCacheError := err.(*RedisCacheError)
+	if err == nil || !isRedisCacheError {
+		t.Fatalf("expecting an error of type RedisCacheError, err=%s", err)
+	}
+
+	//simulate a value that will not expire soon
+	redis.SetTTL(key.String(), 500*time.Second)
+
+	cachedData, err := cache.Get(key)
+	if err != nil {
+		t.Fatalf("failed to get data from redis cache: %s", err)
+	}
+
+	_, err = io.ReadAll(cachedData.Data)
+	_, isRedisCacheError = err.(*RedisCacheError)
+	if err == nil || !isRedisCacheError {
+		t.Fatalf("expecting an error of type RedisCacheError, err=%s", err)
+	}
```

Comment:
Why are we expecting an error here?

Reply:
good question.
In this test, io.ReadAll(cachedData.Data) plays the role of the stream to the http response.
This error could happen in real life if Redis has to clear many items including the one currently streamed.
It could happen, even if the item has an high TTL, if:
- redis is ill sized (way to small to the amount of queries to cache)
- a very big item is put to redis (like an item taken 10% of redis size) while redis is allready full.
These are the only edge cases that can corrupt an http reponse (along with redis being unresponsive during the stream of an item).
Unfortunately, we have no way to "uncorrupt" the stream if we catch this error since the beginning of the stream is already sent. If we replay the query in clickhouse to send the last part of the stream, there is no guarantee clickhouse will give the same result as the previous one partially streamed.
The only perfect way to handle this case  is to always stream redis to a file then stream the file to the http reponse. It has a huge performance drawback. So, I want first to see how often this edge case happens (with the new prometheus metrics added in this PR) to make a decision.

Reply:
Very clear @mga-chka , thank you 🙏